File: app/lua/test/contractAddUser.lua

print('test.contractAddUser')

contractAddUser = {
    --
};

--
local contractFilePath = './../../test/key/crypto/out/c_add_user.crd'; -- contract decrypted

--
local ownerPrikeyFilePath = './../../../../conf/test/key/ed/key_09/ed_privkey.fin';
local ownerPubkeyFilePath = './../../../../conf/test/key/ed/key_09/ed_pubkey.pem';
local superPubkeyFilePath = './../../../../conf/test/key/ed/key_09/ed_pubkey.pem';

--
function contractAddUser.makeContractJson()
    --
    caddUser = cAddUser;

    --
    caddUser:init();

    --------------------------------------------------------
    -- ownerPrikey
    -- io.write('Owner Private key file path to sign (*.fin) : ');
    -- local ownerPrikeyFilePath = io.read();

    -- local ownerPrikey = _commonUtil.readBinaryFile(ownerPrikeyFilePath);
    -- local ownerPrikeyHexStr = _commonUtil.bytesToHexStr(ownerPrikeyHexStr);
    -- print('ownerPrikeyHexStr : ', ownerPrikeyHexStr);

    -- ownerPrikeyPw
    -- io.write('password for *.fin : ');
    -- local ownerPrikeyPw = io.read();
    local ownerPrikeyPw = "asdfQWER1234!@#$";

    --
    local keySeed = ownerPrikeyPw;
    local dec = aesDecFile(ownerPrikeyFilePath, keySeed, string.len(keySeed));
    -- print('dec : ', dec);

    local ownerPrikeyHex = ed25519GetPrikeyByPemStr(dec);
    print('ownerPrikeyHex : ', ownerPrikeyHex);

    -- ownerPubkey
    local ownerPubkey = ed25519GetPubkey(ownerPubkeyFilePath);
    print("ownerPubkey : ", ownerPubkey);
    -- print("ownerPubkey05 : ", _commonUtil.insertStr(ownerPubkey, '05', 0));

    -- superPubkey
    local superPubkey = ed25519GetPubkey(superPubkeyFilePath);
    -- print("superPubkey05 : ", _commonUtil.insertStr(superPubkey, '05', 0));

    --
    local contractActionsJson = _config.contractActions();
    -- print(contractActionsJson['CONTRACT']['DEFAULT']['ADD_USER']);
    -- _commonUtil.prtTable(contractActionsJson['CONTRACT']['DEFAULT']);
    
    --------------------------------------------------------
    --
    local create_tm = utcCurrMS();
    local fintech = '0';
    local privacy = '0';
    local fee = '0';
    local from_account = '0000000000000000';
    local to_account = '0000000000000000';
    local action = contractActionsJson['CONTRACT']['DEFAULT']['ADD_USER'];
    -- contents
    local memo = '';

    --
    caddUser:setContractObj('create_tm', create_tm);
    caddUser:setContractObj('fintech', fintech);
    caddUser:setContractObj('privacy', privacy);
    caddUser:setContractObj('fee', fee);
    caddUser:setContractObj('from_account', from_account);
    caddUser:setContractObj('to_account', to_account);
    caddUser:setContractObj('action', action);
    caddUser:setContractObj('memo', memo);
    
    --------------------------------------------------------
    -- accountId
    local accountId = 'USER_09';

    --
    caddUser:setMyContentObj('owner_pk', _commonUtil.insertStr(ownerPubkey, '05', 0));
    caddUser:setMyContentObj('super_pk', _commonUtil.insertStr(superPubkey, '05', 0));
    caddUser:setMyContentObj('account_id', accountId);

    --
    local mergedBuf = caddUser:signBufferGenerator(true);
    -- print('mergedBuf : ', mergedBuf);

    --
    local inputData = genSha256Str(mergedBuf);
    print('inputData : ', inputData);

    --
    local signature = eddsaSignHex(inputData, ownerPrikeyHex);
    -- print('signature : ', signature);

    --
    caddUser:setContractObj('signed_pubkey', _commonUtil.insertStr(ownerPubkey, '05', 0));
    caddUser:setContractObj('sig', signature);

    -- --
    -- caddUser:prtMyContractObj();
    -- caddUser:prtMyContentObj();
    
    --
    local plaintext = '';
    plaintext = caddUser:jsonStringifyContract(caddUser._contractName, true);
    -- print('plaintext.len : ', string.len(plaintext));
    print('plaintext : ', plaintext);

    _commonUtil.writeBinaryFile(contractFilePath, plaintext);

    --
    caddUser:remove();

    return plaintext;
end

function contractAddUser.testCase()
    --
    local rtnJsonStr = contractAddUser.makeContractJson();
end

File: app/nodejs/test/contractAddUser.js
//
const fs = require('fs');

//
const cryptoSsl = require("./../linux/build/Release/crypto-ssl");

//
const define = require("../njsConn/config/define.js");
// const contractUtil = require("../njsConn/core/contract/contractUtil.js");
const cryptoUtil = require("../njsConn/core/utils/cryptoUtil.js");

//
let contractFilePath = './../../test/key/crypto/out/c_add_user.crd';

//
let ownerPrikeyFilePath = './../../../../conf/test/key/ed/key_09/ed_privkey.fin';
let ownerPubkeyFilePath = './../../../../conf/test/key/ed/key_09/ed_pubkey.pem';
let superPubkeyFilePath = './../../../../conf/test/key/ed/key_09/ed_pubkey.pem';

//
module.exports.makeContentsJson = async () => {
    ////////////////////////////////////////////////////////////////////
    // ownerPrikey
    // let ownerPrikey = fs.readFileSync(ownerPrikeyFilePath, 'binary');

    // ownerPrikeyPw
    let ownerPrikeyPw = "asdfQWER1234!@#$";

    // let stdin = process.openStdin();
    // console.log('password for *.fin : ');
    // stdin.addListener("data", function(d) {
    //     // note:  d is an object, and when converted to a string it will
    //     // end with a linefeed.  so we (rather crudely) account for that  
    //     // with toString() and then trim() 
    //     console.log("you entered: [" +  d.toString().trim() + "]");

    //     ownerPrikeyPw = d.toString().trim();
    // });

    //

    //
    let keySeed = ownerPrikeyPw;
    let dec = cryptoSsl.aesDecFile(ownerPrikeyFilePath, keySeed, keySeed.length);
    console.log('dec : ' + dec);

    let ownerPrikeyHex = cryptoSsl.ed25519GetPrikeyByPemStr(dec);
    console.log('ownerPrikeyHex : ' + ownerPrikeyHex);

    // ownerPubkey
    let ownerPubkey = cryptoSsl.ed25519GetPubkey(ownerPubkeyFilePath);
    console.log("ownerPubkey : " + ownerPubkey);

    // superPubkey
    let superPubkey = cryptoSsl.ed25519GetPubkey(superPubkeyFilePath);
    console.log("superPubkey : " + superPubkey);

    ////////////////////////////////////////////////////////////////////
    //
    let contractJson = {
        create_tm : cryptoSsl.utcCurrMS(), //util.getDateMS().toString(),
        fintech : define.CONTRACT_DEFINE.FINTECH.NON_FINANCIAL_TX,
        privacy : define.CONTRACT_DEFINE.PRIVACY.PUBLIC,
        fee : define.CONTRACT_DEFINE.FEE_DEFAULT,
        from_account : define.CONTRACT_DEFINE.FROM_DEFAULT,
        to_account : define.CONTRACT_DEFINE.TO_DEFAULT,
        action : define.CONTRACT_DEFINE.ACTIONS.DEFAULT.ADD_USER,
        contents : {
            owner_pk : define.CONTRACT_DEFINE.ED_PUB_IDX + ownerPubkey,
            super_pk : define.CONTRACT_DEFINE.ED_PUB_IDX + superPubkey,
            account_id : 'USER_09' // + util.getRandomNumBuf(6).toString('hex')
        },
        memo : ""
    };

    //
    let sig = cryptoUtil.genSign(contractJson, ownerPrikeyHex);
    contractJson.sig = sig;
    contractJson.signed_pubkey = define.CONTRACT_DEFINE.ED_PUB_IDX + ownerPubkey;

    //
    let plaintext = JSON.stringify(contractJson);
    // console.log("plaintext.length : " + plaintext.length);
    console.log("plaintext : " + plaintext);

    fs.writeFileSync(contractFilePath, plaintext, 'binary');
    // fs.writeFileSync(contractFilePath, plaintext);

    return plaintext;
}

module.exports.testCase = async () => {
    this.makeContentsJson();
}
